WordPress/gutenberg
# Convert core toolbar buttons into ToolbarButton

## Context

This PR is part of #18619, whose main goal is to implement [roving tabindex](https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex) on the `@wordpress/components`' `Toolbar` component and to use it on the block toolbars so they become a single tab stop as recommended by the [WAI-ARIA Toolbar Pattern](https://www.w3.org/TR/wai-aria-practices/#toolbar). Related issues are #15331 and #3383.

![GIF showing the navigation through the Gutenberg UI using the tab key](https://user-images.githubusercontent.com/3068563/80681665-825c4000-8a97-11ea-9bd9-0ac570b82591.gif)

This PR improves [`NavigableToolbar`](https://github.com/WordPress/gutenberg/blob/cbcb6195cea11c4dd3469ecfcb54721af873a3a4/packages/block-editor/src/components/navigable-toolbar/index.js) and converts the core toolbar buttons (those buttons that appear on all blocks: movers, block switcher and more options) into [`ToolbarButton`](https://github.com/WordPress/gutenberg/blob/cbcb6195cea11c4dd3469ecfcb54721af873a3a4/packages/components/src/toolbar-button/index.js) or [`ToolbarItem`](https://github.com/WordPress/gutenberg/blob/cbcb6195cea11c4dd3469ecfcb54721af873a3a4/packages/components/src/toolbar-item/index.js).

*`ToolbarItem` is the generic headless component that is used by `ToolbarButton` underneath and can be used to make other components, such as [`DropdownMenu`](https://github.com/WordPress/gutenberg/blob/fd5816e2fed2f241c4eeaea99d94b53af14293db/packages/components/src/dropdown-menu/index.js), aware of the context state provided by the parent [`Toolbar`](https://github.com/WordPress/gutenberg/blob/cbcb6195cea11c4dd3469ecfcb54721af873a3a4/packages/components/src/toolbar/toolbar-container.js) component, which enables the roving tabindex method.*

This also converts a few other buttons so the block toolbar on rich text blocks and other basic blocks already works with roving tabindex.

## `Toolbar` vs. `ToolbarGroup`

> Or `.components-toolbar` vs. `.components-toolbar-group`

Most of the changes on this PR are related to this. On #18534, `Toolbar` was renamed to [`ToolbarGroup`](https://github.com/WordPress/gutenberg/blob/e5dac46fcc0fb44d59d47182a28348a056ab431b/packages/components/src/toolbar-group/index.js), but the old class name (`components-toolbar`) was kept as there's a lot of code relying on it.

However, when `ToolbarGroup` is used within the accessible `Toolbar` component, the class name will be `components-toolbar-group`. That's why it appears duplicated in many parts.

## `NavigableToolbar`

`NavigableToolbar` is the component that is currently used by block toolbars and by the top toolbar. It handles focus when the user presses <kbd>Alt</kbd>+<kbd>F10</kbd>.

With this PR, if all the tabbable DOM elements inside `NavigableToolbar` are rendered by `ToolbarButton` (or `ToolbarItem`), then `NavigableToolbar` will render `Toolbar` instead of [`NavigableMenu`](https://github.com/WordPress/gutenberg/blob/fd5816e2fed2f241c4eeaea99d94b53af14293db/packages/components/src/navigable-container/menu.js). If there's any other tabbable element, it'll work just like it does right now on master.

https://github.com/WordPress/gutenberg/blob/70937da71e8a44a71a82e7814e6eefe7b6104c8e/packages/block-editor/src/components/navigable-toolbar/index.js#L80-L101

This is so we can gradually migrate to this approach, and not break blocks that aren't using `ToolbarButton` for their toolbar buttons, like plugin custom blocks or other blocks like Image (when there's an image). They still work with the <kbd>Tab</kbd> key.

![GIF showing the navigation through the Gutenberg UI using the tab key](https://user-images.githubusercontent.com/3068563/80684954-dae20c00-8a9c-11ea-86f0-d85882f07993.gif)

In short, this code would render `NavigableMenu` because, even though the first item is a `ToolbarButton`, the others are `Button`, `button` and `DropdownMenu`, which aren't aware of `ToolbarContext` and, therefore, can't join the roving tabindex party:

```jsx
<NavigableToolbar aria-label="Block contextual toolbar">
  <ToolbarButton />
  <Button />
  <button />
  <DropdownMenu />
</NavigableToolbar>
```

The code below would render `Toolbar` with roving tabindex enabled:

```jsx
<NavigableToolbar aria-label="Block contextual toolbar">
  <ToolbarButton />
  <ToolbarItem>
    { ( itemProps ) => ( <Button { ...itemProps } /> ) }
  </ToolbarItem>
  <ToolbarItem>
    { ( itemProps ) => ( <button { ...itemProps } /> ) }
  </ToolbarItem>
  <ToolbarItem>
    { ( itemProps ) => ( <DropdownMenu toggleProps={ itemProps } /> ) }
  </ToolbarItem>
</NavigableToolbar>
```

## How to test

- Check rich text blocks and other basic blocks are working with roving tabindex.
- Check if blocks that contain non-`ToolbarItem` buttons are working just like today on master.
- Check if block toolbars are correctly styled, including nested blocks with horizontal movers.
- The top toolbar should work just like on master as the other buttons there haven't been converted yet.

## What's next

- Convert the Top Toolbar buttons.
- Convert toolbar buttons on other blocks.
